File: dart/lib/leetcode/2304/230410.dart
class Solution {
  bool isValid(String s) {
    if (s.length % 2 > 0) {
      return false;
    }
    final open = "([{".codeUnits;
    final close = ")]}".codeUnits;
    List<int> stack = [];

    for (int ch in s.codeUnits) {
      final inOpen = open.indexOf(ch);
      final inClose = close.indexOf(ch);

      if (inClose != -1) {
        if (stack.isEmpty) {
          return false;
        }
        final l = stack.last;
        if (open.indexOf(l) == inClose) {
          stack.removeLast();
        } else {
          return false;
        }
      } else if (inOpen != -1) {
        stack.add(ch);
      }
    }
    return stack.isEmpty;
  }
}

File: rust/230410.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let v: Vec<char> = Vec::new();
        let open = "([{";
        let close = ")]}";

        for i in s.chars() {
            if let posc = close.chars().position(|x| x == i) {
                if let Some(t) = v.pop() {
                    let poso = close.chars().position(|x| x == t);

                    if poso != posc {
                        return false;
                    }
                }
            } else if let poso = open.chars().position(|x| x == i) {
                v.push(i);
            } else {
                // wrong
            }
        }

        v.is_empty()
    }
}
struct Solution {}

fn main() {
    use Solution;
    let sol = Solution::new();
    println!("{}", sol.is_valid("()"));
}
